modular_skyrat/modules/morenarcotics/code/reagents/opium.dm

```diff
@@ -0,0 +1,184 @@
+/datum/chemical_reaction/heroin
+	results = list(/datum/reagent/drug/heroin = 4)
+	required_reagents = list(/datum/reagent/drug/opium = 2, /datum/reagent/acetone = 2)
+	reaction_tags = REACTION_TAG_CHEMICAL
+	required_temp = 480
+	optimal_ph_min = 8
+	optimal_ph_max = 12
+	H_ion_release = -0.04
+	rate_up_lim = 12.5
+	purity_min = 0.5
+
+/datum/chemical_reaction/powder_heroin
+	is_cold_recipe = TRUE
+	required_reagents = list(/datum/reagent/drug/heroin = 8)
+	required_temp = 250 //freeze it
+	reaction_flags = REACTION_INSTANT
+	reaction_flags_skyrat = REACTION_KEEP_INSTANT_REQUIREMENTS
+	reaction_tags = REACTION_TAG_EASY | REACTION_TAG_DRUG | REACTION_TAG_ORGAN | REACTION_TAG_DAMAGING
+	mix_message = "The solution freezes into a powder!"
+
+/datum/chemical_reaction/powder_heroin/on_reaction(datum/reagents/holder, datum/equilibrium/reaction, created_volume)
+	var/location = get_turf(holder.my_atom)
+
+	// get the purity from the holder (can only do this because we arent deleting the ingredients yet)
+	var/saved_purity = 0
+	for(var/used_reagent_type in required_reagents)//this is not an object
+		var/datum/reagent/used_reagent = holder.has_reagent(used_reagent_type)
+		if (!used_reagent)
+			continue
+		saved_purity += used_reagent.creation_purity
+	saved_purity /= required_reagents.len
+
+	// create the result
+	for(var/i in 1 to created_volume)
+		var/obj/item/snortable/heroin/created_heroin = new(location)
+		var/datum/reagent/drug/heroin/powder_heroin/created_reagent = created_heroin.reagents.has_reagent(/datum/reagent/drug/heroin/powder_heroin)
+		created_reagent.creation_purity = saved_purity
+		created_reagent.purity = saved_purity
+		created_heroin.pixel_x = rand(-6, 6)
+		created_heroin.pixel_y = rand(-6, 6)
+
+	// and finally delete the initial ingredients
+	for(var/used_reagent_type in required_reagents)//this is not an object
+		holder.remove_reagent(used_reagent_type, (created_volume * required_reagents[used_reagent_type]), safety = 1)
```

Comment:
I'm so tired of seeing this exact same code copy-pasted everywhere, can't you make this more generic and use subtypes instead?